lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -1614,6 +1621,459 @@ class ConvertAtenAdaptivePoolOp : public OpConversionPattern<OpTy> {
+
+    // left = A00*p00 + A01*p01 = dy0(dx0p00 + dx1p01)
+    Value dx0p00 = b.create<arith::MulFOp>(loc, dx0, p00);
+    Value dx1p01 = b.create<arith::MulFOp>(loc, dx1, p01);
+    Value sum = b.create<arith::AddFOp>(loc, dx0p00, dx1p01);
+    Value left = b.create<arith::MulFOp>(loc, dy0, sum);
+    // right = A10*p10 + A11*p11 = dy1(dx0p10 + dx1p11)
+
+    Value dx0p10 = b.create<arith::MulFOp>(loc, dx0, p10);
+    Value dx1p11 = b.create<arith::MulFOp>(loc, dx1, p11);
+    sum = b.create<arith::AddFOp>(loc, dx0p10, dx1p11);
+    Value right = b.create<arith::MulFOp>(loc, dy1, sum);
+
+    return b.create<arith::AddFOp>(loc, left, right);
+  }
+  LogicalResult
+  matchAndRewrite(Torch::TorchvisionRoiAlignOp op,
+                  typename Torch::TorchvisionRoiAlignOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+
+    Location loc = op->getLoc();
+    Value result = op.getResult();
+
+    uint64_t samplingRatio =
+        cast<ConstantIntOp>(op.getSamplingRatio().getDefiningOp()).getValue();
+    int64_t samplingRatioInt = static_cast<int64_t>(samplingRatio);
+    Value pooledH = op.getPooledHeight();
+    Value pooledW = op.getPooledWidth();
+    Value spatialScaleVal = adaptor.getSpatialScale();
+    llvm::APFloat spatialScale =
+        cast<ConstantFloatOp>(op.getSpatialScale().getDefiningOp()).getValue();
+    Value rois = adaptor.getRois();
+    Value input = adaptor.getInput();
+    RankedTensorType inputType = dyn_cast_or_null<RankedTensorType>(
+        this->getTypeConverter()->convertType(op.getInput().getType()));
+
+    if (inputType == nullptr) {
+      op.emitError("Cannot determine input shape");
+    }
+
+    unsigned inputRank = inputType.getRank();
+    Value offset =
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getF32FloatAttr(0.0));
+    RankedTensorType resultType = dyn_cast_or_null<RankedTensorType>(
+        this->getTypeConverter()->convertType(result.getType()));
+    if (resultType == nullptr) {
+      op.emitError("Cannot determine result shape");
+    }
+
+    Type resultElementType = resultType.getElementType();
+    if (!op.getAligned()) {
+      offset = rewriter.create<arith::ConstantOp>(
+          loc, rewriter.getF32FloatAttr(0.5));
+    }
+
+    Value lb = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value ub0 = rewriter.create<tensor::DimOp>(loc, rois, 0);
+    Value ub1 = rewriter.create<tensor::DimOp>(loc, input, 1);
+    Value step = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    SmallVector<Value> finalOutputShape = {ub0, ub1, pooledH, pooledW};
+    Value finalOutputTensor = rewriter.create<tensor::EmptyOp>(
+        loc, getAsOpFoldResult(finalOutputShape), resultElementType);
+    rewriter.create<scf::ForOp>(
+        loc, lb, ub0, step, ValueRange{},
+        [&](OpBuilder &b, Location loc, Value iv0, ValueRange args) {
```

Comment:
Because this is a long lambda, I'd either make it a free function or at least write out the list of captured variables